File: ios/WingDex/Services/BirdID/EqualEarth.swift
import Foundation

/// Equal Earth (EPSG:8857) projection and grid geometry.
///
/// Port of src/lib/equal-earth.ts. The occurrence prior is keyed by Equal Earth
/// cell, so this decides which cell a photo falls in. A naive lat/lon division
/// would mis-key every lookup, which returns a plausible wrong prior rather
/// than an error, so this stays a straight port and is parity-tested against
/// the TS in BirdIDParityTests.
enum EqualEarth {
    static let gridOriginX = -17_226_000.0
    static let gridOriginY = 8_343_000.0
    static let gridCellSize = 27_000.0
    static let gridCols = 1276
    static let gridRows = 618

    struct Projected: Sendable, Equatable {
        let x: Double
        let y: Double
    }

    struct Cell: Sendable, Equatable {
        let row: Int
        let col: Int
    }

    static func project(lon: Double, lat: Double) -> Projected {
        let a1 = 1.340264, a2 = -0.081106, a3 = 0.000893, a4 = 0.003796
        let a = 6_378_137.0, f = 1.0 / 298.257223563
        let b = a * (1 - f)
        let e2 = 1 - (b * b) / (a * a)
        let e = e2.squareRoot()
        let r = a * (0.5 * (1 + ((1 - e2) / (2 * e)) * log((1 + e) / (1 - e)))).squareRoot()
        let qp = 1 + ((1 - e2) / (2 * e)) * log((1 + e) / (1 - e))
        let lam = lon * .pi / 180
        let phi = lat * .pi / 180
        let sinPhi = sin(phi)
        let eSin = e * sinPhi
        let q = (1 - e2) * (sinPhi / (1 - e2 * sinPhi * sinPhi)
                            - (1 / (2 * e)) * log((1 - eSin) / (1 + eSin)))
        let beta = asin(q / qp)
        let theta = asin((3.0.squareRoot() / 2) * sin(beta))
        let t = theta, t2 = t * t, t6 = t2 * t2 * t2
        let denom = 3 * (a1 + 3 * a2 * t2 + t6 * (7 * a3 + 9 * a4 * t2))
        return Projected(
            x: r * ((2 * 3.0.squareRoot() * lam * cos(t)) / denom),
            y: r * t * (a1 + a2 * t2 + t6 * (a3 + a4 * t2))
        )
    }

    /// Nil when the point falls outside the grid, which is a real case: the
    /// Equal Earth bounding box includes ocean that no cell covers.
    static func cell(x: Double, y: Double) -> Cell? {
        let colD = ((x - gridOriginX) / gridCellSize).rounded(.down)
        let rowD = ((gridOriginY - y) / gridCellSize).rounded(.down)
        guard colD.isFinite, rowD.isFinite else { return nil }
        guard rowD >= 0, rowD < Double(gridRows), colD >= 0, colD < Double(gridCols) else {
            return nil
        }
        return Cell(row: Int(rowD), col: Int(colD))
    }

    static func cell(lat: Double, lon: Double) -> Cell? {
        guard lat.isFinite, lon.isFinite else { return nil }
        let p = project(lon: lon, lat: lat)
        return cell(x: p.x, y: p.y)
    }
}

File: src/lib/equal-earth.ts
// Equal Earth (EPSG:8857) projection and grid geometry.
//
// The occurrence prior is keyed by Equal Earth cell, so this is the projection
// that decides which cell a photo falls in. Pure math: no I/O, no bindings.
//
// Formerly functions/lib/range-adjust.js, which was two modules wearing one
// name. The BirdLife range-trust half was deleted with the GPT path, having
// been measured inert: the fitted smoothing weight landed near 1e-10, putting
// the status term ten orders of magnitude below the occurrence counts. It was
// JavaScript under functions/ because Node build scripts imported it directly
// and there is no allowJs; those scripts are gone, and the only consumers now
// are src/lib/rank.ts and src/lib/occurrence.ts, so it belongs here as TS.

// ── Grid constants (EPSG:8857 Equal Earth) ──────────────────
export const GRID_ORIGIN_X = -17226000
export const GRID_ORIGIN_Y = 8343000
export const GRID_CELL_SIZE = 27000
export const GRID_COLS = 1276
export const GRID_ROWS = 618

// ── Equal Earth projection ──────────────────────────────────

export function lonLatToEqualEarth(lon: number, lat: number): { x: number; y: number } {
  const A1 = 1.340264, A2 = -0.081106, A3 = 0.000893, A4 = 0.003796
  const a = 6378137.0, f = 1 / 298.257223563
  const b = a * (1 - f), e2 = 1 - (b * b) / (a * a), e = Math.sqrt(e2)
  const R = a * Math.sqrt(0.5 * (1 + ((1 - e2) / (2 * e)) * Math.log((1 + e) / (1 - e))))
  const qp = 1 + ((1 - e2) / (2 * e)) * Math.log((1 + e) / (1 - e))
  const lam = (lon * Math.PI) / 180, phi = (lat * Math.PI) / 180
  const sinPhi = Math.sin(phi), eSin = e * sinPhi
  const q = (1 - e2) * (sinPhi / (1 - e2 * sinPhi * sinPhi) - (1 / (2 * e)) * Math.log((1 - eSin) / (1 + eSin)))
  const beta = Math.asin(q / qp)
  const theta = Math.asin((Math.sqrt(3) / 2) * Math.sin(beta))
  const t = theta, t2 = t * t, t6 = t2 * t2 * t2
  const denom = 3 * (A1 + 3 * A2 * t2 + t6 * (7 * A3 + 9 * A4 * t2))
  return {
    x: R * ((2 * Math.sqrt(3) * lam * Math.cos(t)) / denom),
    y: R * t * (A1 + A2 * t2 + t6 * (A3 + A4 * t2)),
  }
}

/** Returns null when the point falls outside the grid, which is a real case:
 *  the Equal Earth bounding box includes ocean that no cell covers. */
export function xyToCell(x: number, y: number): { row: number; col: number } | null {
  const col = Math.floor((x - GRID_ORIGIN_X) / GRID_CELL_SIZE)
  const row = Math.floor((GRID_ORIGIN_Y - y) / GRID_CELL_SIZE)
  if (row < 0 || row >= GRID_ROWS || col < 0 || col >= GRID_COLS) return null
  return { row, col }
}
